Feat(Core): Twitch stream notifications

kyranet opened the pull request:
Fixes #318

Quantumlyy committed:
feat: Twitch notifications (#529)

* feat(Twitch): Challenge handling

* feat(Twitch): Stream update handling

* revert(Util): Didn't think it through

* chore(Twitch): Add config option for Webhook secret

* Did a lot more than what I had planned in this scope

* request(Twitch): Change string literal to const enum

* feat(Twitch): Add SQL table schema

* settings

* remove subscriptions settings

* ci(Twitch): Correct errors

* settings(Twitch): Add types

* settings(Twitch): Add custom message

* refactor(Twitch): Rename raw settings to be prepare for future ones

* feat(Twitch): Add serializer base

Quantumlyy committed:
chore(Twitch): Any can be used instead

Quantumlyy committed:
settings(CommonQuery): Implement methods for Twitch stream subscriptions

Quantumlyy committed:
feat(Twitch): Basic event logic

Quantumlyy committed:
feat(Twitch): Add additional checks for safety

Quantumlyy committed:
Merge branch 'master' into feat-twitch

kyranet committed:
feat: Added PGSQL queries

Quantumlyy committed:
feat(Twitch): Custom embed support

Quantumlyy committed:
Merge remote-tracking branch 'origin/feat-twitch' into feat-twitch

Quantumlyy committed:
refactor(Twitch): Some clean up

Quantumlyy committed:
Merge remote-tracking branch 'origin/master' into feat-twitch

Quantumlyy committed:
feat(Twitch): Ratelimits

PyroTechniac commented on src/lib/util/Notifications/Twitch.ts:
```diff
@@ -0,0 +1,91 @@
+		return result;
+	}
+
+	public static RATELIMIT_COOLDOWN: number = 10 * 60 * 1000;
```
This should use the Time enum found in `lib/util/constants.ts` and multiply `Time.Minute` by 10

kyranet replied:
Resolved with 38cb96f.

kyranet committed:
misc: Refactored events and style changes

kyranet committed:
fix: Resolved CWE-116 vulnerability

kyranet committed:
feat: Added TwitchSubscription serializer

kyranet committed:
fix: Resolved bug where Twitch Stream handlers would reject

The events run asynchronously as they fetch data, and while that is happening, twitchStreamChange has already called response.ok().

When an error happens and the event calls response.error(), it'd throw an "cannot send headers after they were set" error.

This fixes said bug by making the events themselves call `response.ok()` with minimal tweaks so the loop does not return.

Quantumlyy committed:
Start of bearer formation

Quantumlyy committed:
Merge remote-tracking branch 'origin/feat-twitch' into feat-twitch

kyranet committed:
feat: Added skeleton of tsub command

Quantumlyy committed:
feat(Twitch): Bearer token generation

Quantumlyy committed:
Merge remote-tracking branch 'origin/feat-twitch' into feat-twitch

Quantumlyy committed:
fix(Twitch): remove async from fetchBearer

Quantumlyy committed:
feat(Twitch Subscription Command): Allow use of ID or Login

Quantumlyy committed:
feat(Twitch): Subscribe system

kyranet committed:
feat: Handle Twitch subscriptions

kyranet committed:
feat: Finished TODOs and queries

kyranet committed:
misc: Remove CommonQueries#createTwitchStream

kyranet committed:
lint: Resolved ESLint error

kyranet committed:
Merge branch 'master' into feat-twitch

kyranet committed:
i18n: Added missing i18n keys to the language files

kyranet committed:
docs: Fixed typo in TwitchStreamChange#get

kyranet committed:
cleanup: Cleaned up the code from the Twitch class

kyranet committed:
feat: Added TWITCH_CALLBACK to config.ts

kyranet committed:
docs: Documented tsub command

I'm missing two things to do, will do in a moment

kyranet committed:
fix: Handle multi-statuses

kyranet committed:
fix: Resolved build errors

kyranet committed:
misc: Prefix localhost with `http://`

vladfrangu commented on src/commands/Twitch/twitchsubscription.ts:
```diff
@@ -0,0 +1,257 @@
+	Show = 'show'
+}
+
+type Streamer = TwitchHelixUsersSearchResult;
```
Why are all of these aliases a thing?

kyranet replied:
For readability and for shortening the lines in the command arguments, it's hard to read them when they reach +200 characters.

vladfrangu commented on src/lib/util/Notifications/Twitch.ts:
```diff
@@ -0,0 +1,149 @@
+export class Twitch {
+
+	public readonly ratelimitsStreams = new RateLimitManager(1, Twitch.RATELIMIT_COOLDOWN);
+	public readonly BASE_URL_HELIX = 'https://api.twitch.tv/helix/';
```
Shouldn't this use constants?

kyranet replied:
While most of the code uses constants, the newest code has been migrating to moving them as readonly fields, which makes them easier to inspect when debugging or even, to modify (with eval).

This also is, because it makes more sense for them to be class fields when dealing with pure OOP, improving the readability of the code.
